### PR TITLE
Put and query params

### PR DESCRIPTION
**Pull request by jonas27**

### CODE REVIEW

File: server/db.go [hunk 1]
```diff
@@ -4,7 +4,21 @@ type database struct {
...
-func (db *database) get(key string) string {
+func (db *database) keyExists(key string) bool {
 	d := *db.db
-	return d[key]
+	_, ok := d[key]
+	return ok
+}
+
+func (db *database) get(key string) (string, bool) {
+	if !db.keyExists(key) {
```
Comment (Ntr0) [inline server/db.go:14]: this looks like you're reimplementing map value retrieval.

File: server/server.go [hunk 2]
```diff
@@ -37,9 +39,25 @@ func (s *server) handleGet(w http.ResponseWriter, r *http.Request) {
 		log.Println("Error writing response:", err)
 		http.Error(w, "Internal Server Error", http.StatusInternalServerError)
 	}
+}
 
+// according to rfc guidlines PUT should create or replace resources
+// https://www.rfc-editor.org/rfc/rfc2616#section-9.6
+func (s *server) handlePut(w http.ResponseWriter, r *http.Request, key string) {
+	_, ok := s.db.get(key)
```
Comment (Ntr0) [inline server/server.go:47]: keyExists would be semantically better :)

File: server/server.go [hunk 3]
```diff
@@ -37,9 +39,25 @@ func (s *server) handleGet(w http.ResponseWriter, r *http.Request) {
 		log.Println("Error writing response:", err)
 		http.Error(w, "Internal Server Error", http.StatusInternalServerError)
 	}
+}
 
+// according to rfc guidlines PUT should create or replace resources
```
Comment (Ntr0) [inline server/server.go:44]: typo, guidelines

File: server/server.go [hunk 4]
```diff
@@ -37,9 +39,25 @@ func (s *server) handleGet(w http.ResponseWriter, r *http.Request) {
...
+// https://www.rfc-editor.org/rfc/rfc2616#section-9.6
+func (s *server) handlePut(w http.ResponseWriter, r *http.Request, key string) {
+	_, ok := s.db.get(key)
+	if !ok {
+		w.WriteHeader(http.StatusCreated)
```
Comment (Ntr0) [inline server/server.go:49]: You are answering before fulfilling the request

File: server/server_test.go [hunk 5]
```diff
@@ -41,8 +43,48 @@ func TestGet(t *testing.T) {
...
 			is.Equal(w.Code, tt.code)
 			if tt.code == http.StatusOK {
-				is.Equal(w.Body.String(), tt.response)
+				is.Equal(w.Body.String(), tt.respBody)
 			}
 		})
 	}
 }
+
+// TODO: for all tests: split path to path + key vars for better  testing
```
Comment (Ntr0) [inline server/server_test.go:52]: Do the TODO :)

File: server/server_test.go [hunk 6]
```diff
@@ -41,8 +43,48 @@ func TestGet(t *testing.T) {
...
+func TestPut(t *testing.T) {
```
Comment (Ntr0) [inline server/server_test.go:53]: 75% failure. Please fix

File: server/server_test.go [hunk 7]
```diff
@@ -41,8 +43,48 @@ func TestGet(t *testing.T) {
...
+	tests := []struct {
+		name         string
+		path         string
+		reqBody      string
+		overwriteKey bool
+		code         int
+		response     string
+	}{
+		{name: "works, simple", path: "/?key=test", reqBody: "new-entry", code: http.StatusOK, response: "succeeded"},
+		{name: "works, weird body", path: "/?key=test", reqBody: "ntest!@#$%^&*({ }+=)-/\\/test_;'\"", code: http.StatusOK, response: "succeeded"},
```
Comment (Ntr0) [inline server/server_test.go:64]: Weird? That's not even utf-8 :question::exclamation:

**Review by Ntr0 — CHANGES_REQUESTED** [review]

**Comment by jonas27** [comment]
@Ntr0  to avoid too many merge conflicts fixes are here https://github.com/jonas27/ramp-up-go/pull/15

**Comment by jonas27** [comment]
> @Ntr0 to avoid too many merge conflicts fixes are here #15

@Ntr0 @avorima please have a look again :) (PR for fixes is #15 )

**Comment by jonas27** [comment]
> > @Ntr0 to avoid too many merge conflicts fixes are here #15
> 
> @Ntr0 @avorima please have a look again :) (PR for fixes is #15 )

I merge ;)

**Pull request merged**